server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -1885,7 +1894,11 @@ private void doLocalRecovery(
             .<Void>newForked(l -> ActionListener.runWithResource(ActionListener.assertOnce(l), () -> () -> {
                 assert Thread.holdsLock(mutex) == false : "must not hold the mutex here";
                 synchronized (engineMutex) {
-                    IOUtils.close(currentEngineReference.getAndSet(null));
+                    Engine old = currentEngineReference.getAndSet(null);
+                    if (old != null) {
+                        // TODO Remove no-op listener
+                        old.close(ActionListener.noop());
```

Comment:
Likewise here, we're running in an async context already here, we shouldn't need an `ActionListener.noop()`.